cmd/thanos/compact.go

```diff
@@ -326,14 +348,30 @@ func runCompact(
 			level.Warn(logger).Log("msg", "downsampling was explicitly disabled")
 		}
 
-		if err := compact.ApplyRetentionPolicyByResolution(ctx, logger, bkt, metaFetcher, retentionByResolution); err != nil {
+		if err := compact.ApplyRetentionPolicyByResolution(ctx, logger, bkt, metaFetcher, retentionByResolution, blocksMarkedForDeletion); err != nil {
 			return errors.Wrap(err, fmt.Sprintf("retention failed"))
 		}
 
-		compact.BestEffortCleanAbortedPartialUploads(ctx, logger, metaFetcher, bkt, partialUploadDeleteAttempts)
+		compact.BestEffortCleanAbortedPartialUploads(ctx, logger, metaFetcher, bkt, partialUploadDeleteAttempts, blocksMarkedForDeletion)
 		return nil
 	}
 
+	g.Add(func() error {
+		if !wait {
+			return blocksCleaner.DeleteMarkedBlocks(ctx)
+		}
+
+		// --wait=true is specified.
+		return runutil.Repeat(deleteDelay, ctx.Done(), func() error {
```

Comment:
Let's only repeat. The problem with `wait-true` is that now if we do `--wait=false` compactor will probably will delete and quit, with compaction loop being cancelled ):

Reply:
Sure, I have added this. I am not sure if there are any `compact.md` changes required to tell user about this.

Reply:
I am only concerned that since the option name is `wait`, the user will think that if wait is set to false, the process will end after one iteration only. But now the process would not end since we are repeating execution every `deleteDelay` duration. I am unsure if adding notes to `compact.md` about this would be too much. Please share your thoughts.

Reply:
> But now the process would not end since we are repeating execution every deleteDelay duration.

The process will end as soon as first go routine ends, so compaction single run, right?

Reply:
I would ask someone else to look at this, I need to get more knowledge of go routines to comment. @GiedriusS please comment your thoughts, thanks.

Reply:
I'm not completely sure what @bwplotka is exactly suggesting here since this goroutine will continuously try to delete the blocks and it will not end the whole process because `runutil.Repeat` won't return unless it errors out or the context ends. But, I do see a problem with this current code and maybe this is what Bartek meant. `blocksCleaner.DeleteMarkedBlocks()` depends on the meta fetcher having performed at least one sync (successfully). At the beginning of the whole compaction process if `--wait` is `false` then this will surely not execute **after** at least one meta sync in `compactorMainFn()` has happened so it will most likely do nothing unless you are lucky enough. So, we need to ensure this. We could either modify `compactorMainFn()` somehow or we could explicitly perform a meta sync before running the whole group of goroutines. Thoughts? :P

Reply:
It's not about `runutil.Repeat` it's about 2 run groups. Not sure what you are trying to solve here (:

Reply:
As identified in this thread: https://github.com/thanos-io/thanos/pull/2136#discussion_r389825232, I have now added this as part of `compactorMainFn`